tools/jolie2surface/src/joliex/surface/SurfaceCreator.java

```diff
@@ -169,73 +165,96 @@ private String getMax( int max )
+				ret = ret+ "{ ? }";
 			}
-                        else if (type.untypedSubTypes()){ ret = ret+ "{ ? }"; }
+		} else if ( type instanceof TypeChoiceDefinition ){
+			choice = true;
+			ret += getSubType( ( (TypeChoiceDefinition) type).left(), indent ) + " | ";
+			choice = true;
+			ret += getSubType( ( (TypeChoiceDefinition) type ).right(), indent );
+
 		}
-		;
+
 		return ret;
 	}
 
 	private String getType( TypeDefinition type )
 	{
 		String ret = "";                            
-		if ( !types_vector.contains( type.id() ) && !NativeType.isNativeTypeKeyword( type.id() )  && !type.id().equals("undefined")) {
-                        
+		if ( !types_vector.contains( type.id() ) && !NativeType.isNativeTypeKeyword( type.id() )  && !type.id().equals( "undefined" ) ) {
+
 			System.out.print( "type " + type.id() + ":" );
-			if ( type instanceof TypeDefinitionLink ) {
-				System.out.println( ((TypeDefinitionLink) type).linkedTypeName() );
-				if ( !aux_types_vector.contains( ((TypeDefinitionLink) type).linkedType() ) ) {
-					aux_types_vector.add( ((TypeDefinitionLink) type).linkedType() );
+			checkType(type, 1);
+			System.out.println( "" );
+			types_vector.add( type.id() );
+		}
+
+		return ret;
+	}
+
+	private void checkType( TypeDefinition type, int indent ){
+		if ( type instanceof TypeDefinitionLink ) {
+			System.out.print( ( (TypeDefinitionLink) type ).linkedTypeName() );
+			if ( !aux_types_vector.contains( ((TypeDefinitionLink) type).linkedType() ) ) {
+				aux_types_vector.add( ((TypeDefinitionLink) type).linkedType() );
+			}
+		} else  if ( type instanceof TypeInlineDefinition ){
+			System.out.print( ( (TypeInlineDefinition) type ).nativeType().id() );
+			if ( ( (TypeInlineDefinition) type ).hasSubTypes() ) {
+				System.out.print( "{\n" );
+				for( Entry<String, TypeDefinition> entry : ( (TypeInlineDefinition) type ).subTypes() ) {
+					System.out.print( getSubType( entry.getValue(), indent + 1 ) + "\n" );
 				}
+				System.out.print("}");
 			} else {
-				System.out.print( type.nativeType().id() );
-				if ( ((TypeInlineDefinition) type).hasSubTypes() ) {
-					System.out.println( "{" );
-					for( Entry<String, TypeDefinition> entry : ((TypeInlineDefinition) type).subTypes() ) {
-						System.out.println( getSubType( entry.getValue(), 1 ) );
-					}
-					System.out.println( "}" );
-				} else { 
-                                    
-                                    if (type.untypedSubTypes()) {
-					System.out.println( " { ? }" );}
-                                    else {
-                                        System.out.println( "" );
-                                    }
-                                   
-				}
+
+				if ( ( (TypeInlineDefinition) type ).untypedSubTypes() ) {
+					System.out.print( " { ? }" );}
+
 			}
-			types_vector.add( type.id() );
+		} else if ( type instanceof TypeChoiceDefinition ){
+			checkType( ( (TypeChoiceDefinition) type ).left(), 1 );
```

Comment:
Does this way of using static 1 and 3 still work if left() or right() are yet again TypeChoiceDefinition?